### PR TITLE
Report annotation processor details to TAPI progress listeners

### PR DESCRIPTION
**Pull request by marcphilipp**
This PR adds the `JavaCompileTaskSuccessResult` subinterface of
`TaskSuccessResult` which is reported as part of `TaskFinishEvents`.

In order to not add dependencies from :toolingApiBuilder to
:languageJava a new `OperationResultDecoratorFactory` interface is
introduced. It allows to invert the control flow so that plugins can
add additional information by decorating `OperationResult` instances.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/initialization/BuildClientSubscriptions.java [hunk 1]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package org.gradle.tooling.internal.provider;
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/initialization/BuildClientSubscriptions.java]: Why was this moved?

File: subprojects/core/src/main/java/org/gradle/initialization/SubscribableBuildActionRunnerRegistration.java [hunk 2]
```diff
@@ -14,9 +14,8 @@
  * limitations under the License.
  */
 
-package org.gradle.tooling.internal.provider;
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/initialization/SubscribableBuildActionRunnerRegistration.java]: Why was this moved?

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CompileJavaBuildOperationReportingCompiler.java [hunk 3]
```diff
@@ -88,6 +88,18 @@ private DefaultAnnotationProcessorDetails toAnnotationProcessorDetails(Annotatio
     }
 
     private static class Details implements CompileJavaBuildOperationType.Details {
+
+        private final Object taskBuildOperationId;
```
Comment (oehme) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CompileJavaBuildOperationReportingCompiler.java]: Build operations already form a hierarchy, so an observer can already tell that this is the child of a certain Java compile task. So this info is redundant.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultDecoratorFactory.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import org.gradle.tooling.internal.protocol.events.InternalTaskResult;
+import org.gradle.tooling.internal.protocol.events.InternalTaskSuccessResult;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class JavaCompileTaskSuccessResultDecoratorFactory implements OperationResultDecoratorFactory, BuildOperationListener {
```
Comment (oehme) [inline subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultDecoratorFactory.java]: Why just success? We still know how long everything took even if there were compile errors in the end.

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTaskOperationListener.java [hunk 5]
```diff
@@ -92,9 +94,9 @@ public void finished(BuildOperationDescriptor buildOperation, OperationFinishEve
         }
 
         if (buildOperation.getDetails() instanceof ExecuteTaskBuildOperationDetails) {
-            Task task = ((ExecuteTaskBuildOperationDetails) buildOperation.getDetails()).getTask();
-            TaskInternal taskInternal = (TaskInternal) task;
-            eventConsumer.dispatch(new DefaultTaskFinishedProgressEvent(finishEvent.getEndTime(), toTaskDescriptor(buildOperation, taskInternal), toTaskResult(taskInternal, finishEvent)));
+            TaskInternal task = ((ExecuteTaskBuildOperationDetails) buildOperation.getDetails()).getTask();
+            InternalTaskResult result = decoratorFactory.decorate(toTaskResult(task, finishEvent), buildOperation.getId());
```
Comment (oehme) [inline subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTaskOperationListener.java]: Why are we only doing this for tasks?

File: subprojects/tooling-api/tooling-api.gradle.kts [hunk 6]
```diff
@@ -70,6 +70,7 @@ testFixtures {
     from(":logging")
     from(":dependencyManagement")
     from(":ide")
+    from(":languageJava")
```
Comment (oehme) [inline subprojects/tooling-api/tooling-api.gradle.kts]: We should invert this - The test belongs in the java project.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/events/InternalTaskResult.java [hunk 7]
```diff
@@ -22,4 +22,8 @@
  * @since 2.5
  */
 public interface InternalTaskResult extends InternalOperationResult {
+    /**
+     * @since 5.1
+     */
+    String getOutcomeDescription();
```
Comment (oehme) [inline subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/events/InternalTaskResult.java]: Why was this added?

File: subprojects/core/src/main/java/org/gradle/initialization/BuildClientSubscriptions.java (hunk 1 above)
Comment (marcphilipp) [reply subprojects/core/src/main/java/org/gradle/initialization/BuildClientSubscriptions.java]: It was moved to a place where `launcher`, `toolingApiBuilders` and `languageJava` could access it. Do you think there's a better project for it than in `core`? Do you think we should keep `tooling` in the package name?

File: subprojects/core/src/main/java/org/gradle/initialization/SubscribableBuildActionRunnerRegistration.java (hunk 2 above)
Comment (marcphilipp) [reply subprojects/core/src/main/java/org/gradle/initialization/SubscribableBuildActionRunnerRegistration.java]: Ditto.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultDecoratorFactory.java (hunk 4 above)
Comment (marcphilipp) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultDecoratorFactory.java]: I'll change it.

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTaskOperationListener.java (hunk 5 above)
Comment (marcphilipp) [reply subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTaskOperationListener.java]: We only need it for tasks right now and the `decorate()` method only accepts `InternalTaskResult` parameters. I didn't want to add unused methods to the `OperationResultDecoratorFactory`. Or would you like to make the interface more generic and accept `InternalOperationResult` instead of `InternalTaskResult`?

File: subprojects/tooling-api/tooling-api.gradle.kts (hunk 6 above)
Comment (marcphilipp) [reply subprojects/tooling-api/tooling-api.gradle.kts]: Will do.

File: subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/events/InternalTaskResult.java (hunk 7 above)
Comment (marcphilipp) [reply subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/protocol/events/InternalTaskResult.java]: Because `DefaultTaskFinishedProgressEvent` now receives an `InternalTaskResult` instead of an `AbstractTaskResult` and needs to access it in its `getDisplayName()` method.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CompileJavaBuildOperationReportingCompiler.java (hunk 3 above)
Comment (marcphilipp) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CompileJavaBuildOperationReportingCompiler.java]: Done in 9a64bb5cb59.

File: subprojects/tooling-api/tooling-api.gradle.kts (hunk 6 above)
Comment (marcphilipp) [reply subprojects/tooling-api/tooling-api.gradle.kts]: Done in 54eb122399a.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultDecoratorFactory.java (hunk 4 above)
Comment (marcphilipp) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultDecoratorFactory.java]: I tried but we throw an exception instead of returning a result in `JdkJavaCompiler`:

https://github.com/gradle/gradle/blob/c8e18d33be280b0372ef1805822ccbe22bb4beab/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/JdkJavaCompiler.java#L44-L55

**Comment by marcphilipp** [comment]
@oehme PTAL

File: subprojects/core/src/main/java/org/gradle/initialization/BuildClientSubscriptions.java (hunk 1 above)
Comment (oehme) [reply subprojects/core/src/main/java/org/gradle/initialization/BuildClientSubscriptions.java]: I'm not sure I understand why languageJava needs to access this now. Can you point me in the right direction?

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultDecoratorFactory.java (hunk 4 above)
Comment (oehme) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/tooling/JavaCompileTaskSuccessResultDecoratorFactory.java]: I guess we can change this later. People probably don't care much how fast it failed anyway.

File: subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTaskOperationListener.java (hunk 5 above)
Comment (oehme) [reply subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTaskOperationListener.java]: Yeah that's what I was thinking. The decorators can then check for types they care about. I assume you'll need that in the upcoming steps anyway, so we can also do it there.

**Comment by marcphilipp** [comment]
I've moved the protocol event implementations from `:launcher` to `:toolingApi` for now so that both `:launcher` and `:languageJava` can access them. However, they don't belong there. I think we need to introduce a new project, e.g. `:toolingApiProviderBase`.

Moreover, I moved the cross-version test to `:languageJava`. However, now referencing the `HelperProcessorFixture` from the same project [throws a ClassNotFoundException](https://builds.gradle.org/viewLog.html?buildId=17620141&buildTypeId=Gradle_Check_Platform_Java11_Openjdk_Windows_languageJava&tab=buildResultsDiv) at runtime when executed in forking mode.

@oehme Do you know how to fix that?

**Comment by marcphilipp** [comment]
> Why not add a dependency on `:launcher` to `:languageJava`?

That actually worked. Thanks! 🙈

> TAPI tests are loaded in an isolated classloader in order to be able to test different client versions.

FTR, we maintain a list of package filters in `ToolingApiClasspathProvider` and the package from `:languageJava` wasn't included.
